File: Go/echo_client.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"net"
	"os"
	"strings"
)

const (
	HOST = "127.0.0.1"
	PORT = "8888"
)

func main() {
	addr := net.JoinHostPort(HOST, PORT)
	conn, err := net.Dial("tcp", addr)
	if err != nil {
		log.Fatalf("Không kết nối được: %v", err)
	}
	defer conn.Close()

	fmt.Println("Kết nối đến server thành công! Nhập 'exit' để thoát.")

	in := bufio.NewReader(os.Stdin)
	serverReader := bufio.NewReader(conn)

	for {
		fmt.Print("Nhập tin nhắn: ")
		line, err := in.ReadString('\n')
		if err != nil {
			fmt.Printf("Lỗi đọc input: %v\n", err)
			return
		}
		msg := strings.TrimRight(line, "\r\n")
		if strings.EqualFold(strings.TrimSpace(msg), "exit") {
			fmt.Println("Đóng kết nối.")
			return
		}

		// JS echo_client viết msg (không kèm newline). Ta giữ giống.
		if _, err := conn.Write([]byte(msg)); err != nil {
			fmt.Printf("Lỗi gửi: %v\n", err)
			return
		}

		// Đọc phản hồi (echo). Vì server echo raw bytes, có thể không có newline.
		// Ta đọc "tạm" theo buffer.
		buf := make([]byte, 4096)
		n, err := serverReader.Read(buf)
		if n > 0 {
			fmt.Printf("Phản hồi từ server: %s\n", strings.TrimSpace(string(buf[:n])))
		}
		if err != nil {
			fmt.Printf("Server đã đóng kết nối (%v)\n", err)
			return
		}
	}
}

File: Node.js/echo_client.js
// echo_client.js
const net = require("net");
const readline = require("readline");

const HOST = "127.0.0.1";
const PORT = 8888;

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const socket = net.createConnection({ host: HOST, port: PORT }, () => {
  console.log("Kết nối đến server thành công! Nhập 'exit' để thoát.");
  prompt();
});

socket.on("data", (chunk) => {
  console.log(`Phản hồi từ server: ${chunk.toString("utf8").trim()}`);
  prompt();
});

socket.on("end", () => {
  console.log("Server đã đóng kết nối.");
  rl.close();
});

socket.on("error", (err) => {
  console.log(`Lỗi: ${err.message}`);
  rl.close();
});

function prompt() {
  rl.question("Nhập tin nhắn: ", (msg) => {
    if (msg.trim().toLowerCase() === "exit") {
      console.log("Đóng kết nối.");
      socket.end();
      rl.close();
      return;
    }
    socket.write(msg);
  });
}
